cpp/oneapi/dal/test/engine/dataframe_common.hpp

```diff
@@ -0,0 +1,241 @@
+
+#pragma once
+
+#include <any>
+#include <string>
+#include <vector>
+#include <memory>
+#include <optional>
+
+#include "oneapi/dal/array.hpp"
+#include "oneapi/dal/table/common.hpp"
+#include "oneapi/dal/test/engine/common.hpp"
+
+namespace oneapi::dal::test::engine {
+
+#define GENERATE_DATAFRAME(...) \
+    GENERATE(as<oneapi::dal::test::engine::dataframe_builder>{}, __VA_ARGS__).build()
+
+class dataframe_impl {
+public:
+    explicit dataframe_impl(const array<float>& data,
+                            std::int64_t row_count,
+                            std::int64_t column_count)
+            : array_(data),
+              row_count_(row_count),
+              column_count_(column_count) {
+        array_.need_mutable_data();
+    }
+
+    dataframe_impl(const dataframe_impl&) = delete;
+    dataframe_impl& operator=(const dataframe_impl&) = delete;
+
+    std::int64_t get_count() const {
+        return array_.get_count();
+    }
+
+    std::size_t get_size() const {
+        return array_.get_size();
+    }
+
+    std::int64_t get_row_count() const {
+        return row_count_;
+    }
+
+    std::int64_t get_column_count() const {
+        return column_count_;
+    }
+
+    float* get_data() const {
+        return array_.get_mutable_data();
+    }
+
+    const array<float>& get_array() const {
+        return array_;
+    }
+
+    dataframe_impl* copy() const {
+        auto array_copy = array<float>::empty(array_.get_count());
+        float* array_copy_data = array_copy.get_mutable_data();
+        const float* array_data = array_.get_data();
+        for (std::int64_t i = 0; i < array_.get_count(); i++) {
+            array_copy_data[i] = array_data[i];
+        }
+        return new dataframe_impl{ array_copy, row_count_, column_count_ };
+    }
+
+    template <typename T>
+    void add_field(const std::string& name, T&& value) {
+        user_fields_.insert_or_assign(name, std::any{ std::forward<T>(value) });
+    }
+
+    void remove_field(const std::string& name) {
+        user_fields_.erase(name);
+    }
+
+    template <typename T>
+    std::optional<T> get_field(const std::string& name) const {
+        const auto it = user_fields_.find(name);
+        if (it == user_fields_.end()) {
+            return std::nullopt;
+        }
+
+        try {
+            return std::any_cast<T>(it->second);
+        }
+        catch (const std::bad_any_cast&) {
+            return std::nullopt;
+        }
+    }
+
+private:
+    array<float> array_;
+    std::int64_t row_count_;
+    std::int64_t column_count_;
+    std::unordered_map<std::string, std::any> user_fields_;
+};
+
+class dataframe {
+public:
+    explicit dataframe(const array<float>& data, std::int64_t row_count, std::int64_t column_count)
+            : dataframe(new dataframe_impl{ data, row_count, column_count }) {}
+
+    explicit dataframe(dataframe_impl* impl) : impl_(impl) {}
+
+    template <typename Float>
```

Comment:
I don't see that template parameter is used in implementation

Reply:
Yes, it should be passed to table builder

Reply:
Fixed